src/thread_event.c

```diff
@@ -0,0 +1,254 @@
+			wait = event_wait;				\
+		}							\
+	}
+
+	ITERATE_OVER_ALL_EVENTS
+#undef E
+
+	assert(no_event_on == !thread_event_active);
+	assert(wait <= THREAD_EVENT_MAX_START_WAIT);
+	return wait;
+}
+
+void
+thread_event_assert_invariants_debug(tsd_t *tsd) {
+	uint64_t thread_allocated = thread_allocated_get(tsd);
+	uint64_t last_event = thread_allocated_last_event_get(tsd);
+	uint64_t next_event = thread_allocated_next_event_get(tsd);
+	uint64_t next_event_fast = thread_allocated_next_event_fast_get(tsd);
+
+	assert(last_event != next_event);
+	if (next_event <= THREAD_ALLOCATED_NEXT_EVENT_FAST_MAX) {
+		assert(next_event_fast == next_event);
+	} else {
+		assert(next_event_fast == 0U);
+	}
+
+	/* The subtraction is intentionally susceptible to underflow. */
+	uint64_t interval = next_event - last_event;
+
+	/* The subtraction is intentionally susceptible to underflow. */
+	assert(thread_allocated - last_event < interval);
+
+	uint64_t min_wait = thread_allocated_next_event_compute(tsd);
+
+	/*
+	 * next_event should have been pushed up only except when no event is
+	 * on and the TSD is just initialized.  The last_event == 0U guard
+	 * below is stronger than needed, but having an exactly accurate guard
+	 * is more complicated to implement.
+	 */
+	assert((!thread_event_active && last_event == 0U) ||
+	    interval == min_wait ||
+	    (interval < min_wait && interval == THREAD_EVENT_MAX_INTERVAL));
+}
+
+static void
+thread_event_adjust_thresholds_helper(tsd_t *tsd, uint64_t last_event,
+    uint64_t wait) {
+	assert(last_event == thread_allocated_last_event_get(tsd));
+	assert(wait <= THREAD_EVENT_MAX_START_WAIT);
+	uint64_t next_event = last_event + (wait <= THREAD_EVENT_MAX_INTERVAL ?
+	    wait : THREAD_EVENT_MAX_INTERVAL);
+	thread_allocated_next_event_set(tsd, next_event);
+	uint64_t next_event_fast = (next_event <=
+	    THREAD_ALLOCATED_NEXT_EVENT_FAST_MAX) ? next_event : 0U;
+	thread_allocated_next_event_fast_set(tsd, next_event_fast);
+}
+
+static void
+thread_prof_sample_event_handler(tsd_t *tsd) {
+	assert(config_prof && opt_prof);
+	assert(prof_sample_event_wait_get(tsd) == 0U);
+	if (!prof_active_get_unlocked()) {
+		/*
+		 * If prof_active is off, we reset prof_sample_event_wait to be
+		 * the sample interval when it drops to 0, so that there won't
+		 * be excessive routings to the slow path, and that when
+		 * prof_active is turned on later, the counting for sampling
+		 * can immediately resume as normal.
+		 */
+		thread_prof_sample_event_update(tsd,
+		    (uint64_t)(1 << lg_prof_sample));
+	}
+}
+
+void
+thread_event_trigger(tsd_t *tsd, uint64_t usize, bool delay_event) {
+	/* usize has already been added to thread_allocated. */
+	uint64_t thread_allocated_after = thread_allocated_get(tsd);
+
+	/* The subtraction is intentionally susceptible to underflow. */
+	uint64_t accumbytes = thread_allocated_after -
+	    thread_allocated_last_event_get(tsd);
+
+	if (accumbytes < usize) {
```

Comment:
With the max_interval, it seems this cannot overflow anymore? I'd say turn this into an assert is fine; up to you.

Reply:
It is still possible, if `thread_allocated` has already advanced significantly before this allocation, and this allocation is large enough so that the total allocation since last event overflows `uint64_t`.

Reply:
Hmm I mean with the max interval, it should always go down the trigger path once in a while, meaning thread.allocated cannot advance significantly without updating last event?

Reply:
e.g. this should be true: cassert(UINT64_MAX - max single alloc size > MAX_INTERVAL)

Reply:
Right, if `cassert(UINT64_MAX - SC_LARGE_MAXCLASS >= MAX_INTERVAL - 1)` holds then `accumbytes` won't overflow. But it doesn't seem obvious that the `cassert()` condition is guaranteed? I guess I can define `MAX_INTERVAL` to be the smaller of `4 << 20` and `UINT64_MAX - SC_LARGE_MAXCLASS + 1` to be rigorous?

Reply:
No strongly preference here. Slightly prefer leaving the max_interval macro alone, and add a cassert about max interval + large_maxclass <= uint64max.

Reply:
Will add a `cassert()`, and will also remove the `usize` parameter for `thread_event_trigger()`.